substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+						"bonded and active validator should also be part of the voter list"
+					);
+					// return self-stake (ie. active bonded).
+					T::Staking::stake(&target).map(|s| Self::weight_of(s.active)).ok()
+				},
+				Ok(StakerStatus::Nominator(_)) => {
+					panic!("staker with nominator status should not be part of the target list");
+				},
+			};
+
+			if let Some(score) = maybe_self_stake {
+				if let Some(stake) = approvals_map.get_mut(&target) {
+					*stake += score as ExtendedBalance;
+				} else {
+					approvals_map.insert(target, score.into());
+				}
+			} else {
+				// unbonded target: it does not have self-stake.
+			}
+		}
+
+		log!(trace, "try-state: calculated approvals map: {:?}", approvals_map);
+
+		// compare calculated approvals per target with target list state.
+		for (target, calculated_stake) in approvals_map.iter() {
+			let stake_in_list = T::TargetList::get_score(target).expect("target must exist; qed.");
+			let stake_in_list = Self::to_vote_extended(stake_in_list);
+
+			if *calculated_stake != stake_in_list {
+				log!(
+						error,
+						"try-runtime: score of {:?} in `TargetList` list: {:?}, calculated sum of all stake: {:?}",
+						target,
+						stake_in_list,
+						calculated_stake,
+					);
+
+				return Err("target score in the target list is different than the expected".into())
+			}
+		}
+
+		frame_support::ensure!(
+			approvals_map.keys().count() == T::TargetList::iter().count(),
+			"calculated approvals count is different from total of target list.",
+		);
+
+		Ok(())
+	}
+
+	/// Try-state: checks if targets in the target list are sorted by score.
+	///
+	/// Invariant
+	///  * All targets in the target list are sorted by their score.
+	///
+	///  NOTE: unfortunatelly, it is not trivial to check if the sort correctness of the list if
+	///  the `SortedListProvider` is implemented by bags list due to score bucketing. Thus, we
+	///  leverage the [`SortedListProvider::in_position`] to verify if the target is in the
+	/// correct  position in the list (bag or otherwise), given its score.
+	#[cfg(feature = "try-runtime")]
+	pub fn do_try_state_target_sorting() -> Result<(), sp_runtime::TryRuntimeError> {
+		for t in T::TargetList::iter() {
+			frame_support::ensure!(
+				T::TargetList::in_position(&t).expect("target exists"),
+				"target list is not sorted"
+			);
+		}
+
+		Ok(())
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// Fired when the stake amount of some staker updates.
+	///
+	/// When a nominator's stake is updated, all the nominated targets must be updated
+	/// accordingly.
+	///
+	/// Note: it is assumed that `who`'s staking ledger state is updated *before* this method is
+	/// called.
+	fn on_stake_update(
+		who: &T::AccountId,
+		prev_stake: Option<Stake<BalanceOf<T>>>,
+		stake: Stake<BalanceOf<T>>,
+	) {
+		// closure to calculate the stake imbalance of a staker.
+		let stake_imbalance_of = |prev_stake: Option<Stake<BalanceOf<T>>>,
```

Comment:
maybe it could be cleanner if this is a free-standing fn, or a constructor on the type `StakeImbalance`